File: app/store.py
"""In-memory booking store. Swap for a real database when you deploy for real."""

from __future__ import annotations

import json
import random
import string
from datetime import date, datetime, timedelta

SERVICES = {
    "haircut": {
        "label": "Signature Haircut & Styling",
        "minutes": 45,
        "price": 85,
        "description": "Custom consultation, precision shear or razor cut, wash, and luxury blowout.",
    },
    "coloring": {
        "label": "Full Color & Gloss",
        "minutes": 90,
        "price": 185,
        "description": "All-over single process coloring, custom formulation, nourishing gloss, and blowout.",
    },
    "balayage": {
        "label": "Artisan Balayage & Highlights",
        "minutes": 120,
        "price": 280,
        "description": "Hand-painted dimensional highlights, toner formulation, deep conditioning mask, and style.",
    },
    "blowout": {
        "label": "Signature Blowout & Treatment",
        "minutes": 45,
        "price": 65,
        "description": "Revitalizing scalp massage, clarifying shampoo, hydrating mask, and voluminous blowout styling.",
    },
}

OPEN_HOUR = 9
CLOSE_HOUR = 17
SLOT_MINUTES = 30
HORIZON_DAYS = 14

_appointments: dict[str, dict] = {}
_taken: set[tuple[str, str]] = set()


def _slots_for_day(day: date) -> list[str]:
    start = datetime.combine(day, datetime.min.time()).replace(hour=OPEN_HOUR)
    end = datetime.combine(day, datetime.min.time()).replace(hour=CLOSE_HOUR)
    slots: list[str] = []
    cursor = start
    while cursor < end:
        slots.append(cursor.strftime("%H:%M"))
        cursor += timedelta(minutes=SLOT_MINUTES)
    return slots


def _seed_existing_bookings() -> None:
    """Pre-book a deterministic scatter of slots so availability looks real."""
    rng = random.Random(20260901)
    for day_offset in range(HORIZON_DAYS):
        day = date.today() + timedelta(days=day_offset)
        if day.weekday() >= 5:
            continue
        for slot in _slots_for_day(day):
            if rng.random() < 0.45:
                _taken.add((day.isoformat(), slot))

    # Seed 4 realistic appointments for the Owner Dashboard
    seed_data = [
        ("haircut", 0, "10:00", "Eleanor Vance", "eleanor.vance@gmail.com", True),
        ("blowout", 0, "14:30", "Marcus Sterling", "marcus.sterling@gmail.com", True),
        ("coloring", 1, "11:00", "Sophia Al-Mansoor", "sophia.mansoor@gmail.com", False),
        ("balayage", 2, "09:30", "James Thornton", "james.thornton@gmail.com", True),
    ]
    for service, offset, time_str, name, email, confirmed in seed_data:
        target_day = date.today() + timedelta(days=offset)
        if target_day.weekday() >= 5:
            target_day += timedelta(days=2)
        code = "".join(random.choices(string.ascii_uppercase + string.digits, k=6))
        _taken.add((target_day.isoformat(), time_str))
        _appointments[code] = {
            "confirmation_code": code,
            "service": service,
            "service_label": SERVICES[service]["label"],
            "date": target_day.isoformat(),
            "time": time_str,
            "customer_name": name,
            "email": email,
            "confirmation_sent": confirmed,
            "price": SERVICES[service]["price"],
            "created_at": datetime.now().strftime("%Y-%m-%d %H:%M"),
        }


def _confirmation_code() -> str:
    alphabet = string.ascii_uppercase + string.digits
    while True:
        code = "".join(random.choices(alphabet, k=6))
        if code not in _appointments:
            return code


def is_open(day: date) -> bool:
    return day.weekday() < 5


def available_slots(day: date, limit: int = 4) -> list[str]:
    if not is_open(day):
        return []
    free = [s for s in _slots_for_day(day) if (day.isoformat(), s) not in _taken]
    return free[:limit]


def next_open_days(after: date, count: int = 2) -> list[date]:
    found: list[date] = []
    cursor = after
    while len(found) < count:
        cursor += timedelta(days=1)
        if (cursor - date.today()).days > HORIZON_DAYS:
            break
        if is_open(cursor) and available_slots(cursor):
            found.append(cursor)
    return found


def book(service: str, day: date, time_str: str, name: str, email: str) -> dict:
    key = (day.isoformat(), time_str)
    if key in _taken:
        raise SlotUnavailable(f"{time_str} on {day.isoformat()} is already booked")
    _taken.add(key)
    code = _confirmation_code()
    record = {
        "confirmation_code": code,
        "service": service,
        "service_label": SERVICES[service]["label"],
        "date": day.isoformat(),
        "time": time_str,
        "customer_name": name,
        "email": email,
        "confirmation_sent": False,
        "price": SERVICES[service].get("price", 120),
        "created_at": datetime.now().strftime("%Y-%m-%d %H:%M"),
    }
    _appointments[code] = record
    return record


def get(code: str) -> dict | None:
    return _appointments.get(code)


def mark_confirmation_sent(code: str) -> dict | None:
    record = _appointments.get(code)
    if record:
        record["confirmation_sent"] = True
    return record


class SlotUnavailable(Exception):
    pass


_seed_existing_bookings()


# --- demo event log -------------------------------------------------------
# HTTP tools run on AssemblyAI's servers, so the browser never sees a
# tool.call event. Recording each hit here is what lets the demo page show
# them arriving.

_events: list[dict] = []


def _safe_json(raw: bytes) -> dict:
    try:
        return json.loads(raw or b"{}")
    except (ValueError, TypeError):
        return {}


def log_event(path: str, request_body: bytes, response_body: bytes) -> None:
    _events.append({
        "seq": len(_events) + 1,
        "tool": path.rsplit("/", 1)[-1],
        "arguments": _safe_json(request_body),
        "result": _safe_json(response_body),
        "at": datetime.now().strftime("%H:%M:%S"),
    })


def events_since(cursor: int) -> list[dict]:
    return [e for e in _events if e["seq"] > cursor]

File: src/lib/store.ts
/**
 * In-memory booking store matching legacy store.py.
 * Dual DB Layer: Works alongside persistent SQLite database (src/lib/db.ts).
 */

export interface Appointment {
  confirmation_code: string;
  service: string;
  service_label: string;
  date: string;
  time: string;
  customer_name: string;
  email: string;
  confirmation_sent: boolean;
  price: number;
  created_at: string;
}

export interface ToolEvent {
  seq: number;
  tool: string;
  arguments: any;
  result: any;
  at: string;
}

export const SERVICES: Record<
  string,
  { label: string; minutes: number; price: number; description: string }
> = {
  haircut: {
    label: "Signature Haircut & Styling",
    minutes: 45,
    price: 85,
    description:
      "Custom consultation, precision shear or razor cut, wash, and luxury blowout.",
  },
  coloring: {
    label: "Full Color & Gloss",
    minutes: 90,
    price: 185,
    description:
      "All-over single process coloring, custom formulation, nourishing gloss, and blowout.",
  },
  balayage: {
    label: "Artisan Balayage & Highlights",
    minutes: 120,
    price: 280,
    description:
      "Hand-painted dimensional highlights, toner formulation, deep conditioning mask, and style.",
  },
  blowout: {
    label: "Signature Blowout & Treatment",
    minutes: 45,
    price: 65,
    description:
      "Revitalizing scalp massage, clarifying shampoo, hydrating mask, and voluminous blowout styling.",
  },
};

interface StoreState {
  appointments: Record<string, Appointment>;
  taken: Set<string>; // "YYYY-MM-DD|HH:MM"
  events: ToolEvent[];
  openHour: number;
  closeHour: number;
  slotMinutes: number;
  services: typeof SERVICES;
}

function getSlotsForDay(openH = 9, closeH = 17, step = 30): string[] {
  const slots: string[] = [];
  let cur = openH * 60;
  const end = closeH * 60;
  while (cur < end) {
    const h = Math.floor(cur / 60);
    const m = cur % 60;
    slots.push(`${String(h).padStart(2, "0")}:${String(m).padStart(2, "0")}`);
    cur += step;
  }
  return slots;
}

function generateConfirmationCode(state: StoreState): string {
  const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
  while (true) {
    let code = "";
    for (let i = 0; i < 6; i++) {
      code += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    if (!state.appointments[code]) {
      return code;
    }
  }
}

function pseudoRandom(seed: number) {
  let s = seed;
  return function () {
    s = (s * 1664525 + 1013904223) % 4294967296;
    return s / 4294967296;
  };
}

function initStoreState(): StoreState {
  const state: StoreState = {
    appointments: {},
    taken: new Set<string>(),
    events: [],
    openHour: 9,
    closeHour: 17,
    slotMinutes: 30,
    services: { ...SERVICES },
  };

  const rng = pseudoRandom(20260901);
  const today = new Date();
  const pad = (n: number) => String(n).padStart(2, "0");

  // Scatter 45% over 14 days
  for (let dayOffset = 0; dayOffset < 14; dayOffset++) {
    const d = new Date(today);
    d.setDate(today.getDate() + dayOffset);
    const dayOfWeek = d.getDay(); // 0 Sun, 6 Sat
    if (dayOfWeek === 0 || dayOfWeek === 6) continue;
    const iso = `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`;

    const slots = getSlotsForDay(state.openHour, state.closeHour, state.slotMinutes);
    for (const slot of slots) {
      if (rng() < 0.45) {
        state.taken.add(`${iso}|${slot}`);
      }
    }
  }

  // 4 seed appointments for owner dashboard
  const seedData = [
    {
      service: "haircut",
      offset: 0,
      timeStr: "10:00",
      name: "Eleanor Vance",
      email: "eleanor.vance@gmail.com",
      confirmed: true,
    },
    {
      service: "blowout",
      offset: 0,
      timeStr: "14:30",
      name: "Marcus Sterling",
      email: "marcus.sterling@gmail.com",
      confirmed: true,
    },
    {
      service: "coloring",
      offset: 1,
      timeStr: "11:00",
      name: "Sophia Al-Mansoor",
      email: "sophia.mansoor@gmail.com",
      confirmed: false,
    },
    {
      service: "balayage",
      offset: 2,
      timeStr: "09:30",
      name: "James Thornton",
      email: "james.thornton@gmail.com",
      confirmed: true,
    },
  ];

  for (const item of seedData) {
    const d = new Date(today);
    d.setDate(today.getDate() + item.offset);
    if (d.getDay() === 0) d.setDate(d.getDate() + 1);
    if (d.getDay() === 6) d.setDate(d.getDate() + 2);
    const iso = `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`;

    const code = generateConfirmationCode(state);
    state.taken.add(`${iso}|${item.timeStr}`);
    const srv = state.services[item.service] || SERVICES[item.service];
    state.appointments[code] = {
      confirmation_code: code,
      service: item.service,
      service_label: srv?.label || item.service,
      date: iso,
      time: item.timeStr,
      customer_name: item.name,
      email: item.email,
      confirmation_sent: item.confirmed,
      price: srv?.price || 85,
      created_at: new Date().toISOString().replace("T", " ").slice(0, 16),
    };
  }

  return state;
}

// Attach to globalThis so state persists across Next.js dev server hot reloads
const globalStore = (globalThis as any).__omnidesk_store || initStoreState();
if (process.env.NODE_ENV !== "production") {
  (globalThis as any).__omnidesk_store = globalStore;
}

export const store = {
  get state(): StoreState {
    return globalStore;
  },

  isOpen(dateStr: string): boolean {
    const d = new Date(dateStr + "T00:00:00Z");
    const day = d.getUTCDay();
    return day >= 1 && day <= 5;
  },

  availableSlots(dateStr: string, limit = 4): string[] {
    if (!this.isOpen(dateStr)) return [];
    const all = getSlotsForDay(
      globalStore.openHour,
      globalStore.closeHour,
      globalStore.slotMinutes
    );
    const free = all.filter((s) => !globalStore.taken.has(`${dateStr}|${s}`));
    return free.slice(0, limit);
  },

  isSlotTaken(dateStr: string, timeStr: string): boolean {
    return globalStore.taken.has(`${dateStr}|${timeStr}`);
  },

  nextOpenDays(afterDateStr: string, count = 2): string[] {
    const found: string[] = [];
    const cur = new Date(afterDateStr + "T00:00:00Z");
    const today = new Date();
    const pad = (n: number) => String(n).padStart(2, "0");

    for (let i = 0; i < 14 && found.length < count; i++) {
      cur.setUTCDate(cur.getUTCDate() + 1);
      const iso = `${cur.getUTCFullYear()}-${pad(cur.getUTCMonth() + 1)}-${pad(
        cur.getUTCDate()
      )}`;
      if (this.isOpen(iso) && this.availableSlots(iso, 1).length > 0) {
        found.push(iso);
      }
    }
    return found;
  },

  book(
    service: string,
    dateStr: string,
    timeStr: string,
    name: string,
    email: string
  ): Appointment {
    const key = `${dateStr}|${timeStr}`;
    if (globalStore.taken.has(key)) {
      throw new Error(`${timeStr} on ${dateStr} is already booked`);
    }
    globalStore.taken.add(key);
    const code = generateConfirmationCode(globalStore);
    const srv = globalStore.services[service] || SERVICES[service];
    const record: Appointment = {
      confirmation_code: code,
      service,
      service_label: srv?.label || service,
      date: dateStr,
      time: timeStr,
      customer_name: name,
      email,
      confirmation_sent: false,
      price: srv?.price ?? 120,
      created_at: new Date().toISOString().replace("T", " ").slice(0, 16),
    };
    globalStore.appointments[code] = record;
    return record;
  },

  get(code: string): Appointment | undefined {
    return globalStore.appointments[code.trim()];
  },

  markConfirmationSent(code: string): Appointment | undefined {
    const record = globalStore.appointments[code.trim()];
    if (record) {
      record.confirmation_sent = true;
    }
    return record;
  },

  getAllAppointments(): Appointment[] {
    return Object.values(globalStore.appointments);
  },

  logEvent(path: string, requestBody: any, responseBody: any): void {
    const toolName = path.split("/").filter(Boolean).pop() || "unknown";
    const now = new Date();
    const timeStr = `${String(now.getHours()).padStart(2, "0")}:${String(
      now.getMinutes()
    ).padStart(2, "0")}:${String(now.getSeconds()).padStart(2, "0")}`;

    const event: ToolEvent = {
      seq: globalStore.events.length + 1,
      tool: toolName,
      arguments: requestBody && typeof requestBody === "object" ? requestBody : {},
      result: responseBody && typeof responseBody === "object" ? responseBody : {},
      at: timeStr,
    };
    globalStore.events.push(event);
  },

  eventsSince(cursor: number): ToolEvent[] {
    return globalStore.events.filter((e: ToolEvent) => e.seq > cursor);
  },
};
